From d1102c663c8ed9dcef140c7eba755386e8746407 Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Date: Thu, 27 Nov 2014 04:03:15 +0200
Subject: [PATCH 317/456] v4l: xilinx: resample: Use xvip_init_resources()

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

Replace the manual ioremap by a call to xvip_init_resources(), which
also handles the functional clock.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
(cherry picked from commit e6d297de00cba86421950a5f82d183da3155bbb4)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 .../bindings/media/xilinx/xlnx,v-cresample.txt     |    3 +++
 drivers/media/platform/xilinx/xilinx-cresample.c   |   13 +++++++------
 2 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/Documentation/devicetree/bindings/media/xilinx/xlnx,v-cresample.txt b/Documentation/devicetree/bindings/media/xilinx/xlnx,v-cresample.txt
index 471c1f8..819db78 100644
--- a/Documentation/devicetree/bindings/media/xilinx/xlnx,v-cresample.txt
+++ b/Documentation/devicetree/bindings/media/xilinx/xlnx,v-cresample.txt
@@ -7,6 +7,8 @@ Required properties:
 
 - reg: Physical base address and length of the registers set for the device.
 
+- clocks: Reference to the video core clock.
+
 - ports: Video ports, using the DT bindings defined in ../video-interfaces.txt.
   The cresample as han input port (0) and an output port (1).
 
@@ -22,6 +24,7 @@ Example:
 	axi_cresample_0: axi_cresample {
 		compatible = "xlnx,v-cresample-4.0";
 		reg = <0x40120000 0x10000>;
+		clocks = <&clkc 15>;
 
 		ports {
 			#address-cells = <1>;
diff --git a/drivers/media/platform/xilinx/xilinx-cresample.c b/drivers/media/platform/xilinx/xilinx-cresample.c
index 8c06b9c..1da61ef 100644
--- a/drivers/media/platform/xilinx/xilinx-cresample.c
+++ b/drivers/media/platform/xilinx/xilinx-cresample.c
@@ -316,7 +316,6 @@ static int xcresample_parse_of(struct xcresample_device *xcresample)
 static int xcresample_probe(struct platform_device *pdev)
 {
 	struct xcresample_device *xcresample;
-	struct resource *res;
 	struct v4l2_subdev *subdev;
 	struct v4l2_mbus_framefmt *default_format;
 	int ret;
@@ -331,10 +330,9 @@ static int xcresample_probe(struct platform_device *pdev)
 	if (ret < 0)
 		return ret;
 
-	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
-	xcresample->xvip.iomem = devm_ioremap_resource(&pdev->dev, res);
-	if (IS_ERR(xcresample->xvip.iomem))
-		return PTR_ERR(xcresample->xvip.iomem);
+	ret = xvip_init_resources(&xcresample->xvip);
+	if (ret < 0)
+		return ret;
 
 	/* Reset and initialize the core */
 	xvip_reset(&xcresample->xvip);
@@ -368,7 +366,7 @@ static int xcresample_probe(struct platform_device *pdev)
 	subdev->entity.ops = &xcresample_media_ops;
 	ret = media_entity_init(&subdev->entity, 2, xcresample->pads, 0);
 	if (ret < 0)
-		return ret;
+		goto error;
 
 	v4l2_ctrl_handler_init(&xcresample->ctrl_handler, 2);
 	xcresample_field.def =
@@ -403,6 +401,7 @@ static int xcresample_probe(struct platform_device *pdev)
 error:
 	v4l2_ctrl_handler_free(&xcresample->ctrl_handler);
 	media_entity_cleanup(&subdev->entity);
+	xvip_cleanup_resources(&xcresample->xvip);
 	return ret;
 }
 
@@ -415,6 +414,8 @@ static int xcresample_remove(struct platform_device *pdev)
 	v4l2_ctrl_handler_free(&xcresample->ctrl_handler);
 	media_entity_cleanup(&subdev->entity);
 
+	xvip_cleanup_resources(&xcresample->xvip);
+
 	return 0;
 }
 
-- 
1.7.5.4

